File: trunk/FftGuitarTuner/FrequenciesScale.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Drawing;

namespace FftGuitarTuner
{
    public partial class FrequenciesScale : UserControl
    {
        const double MinFrequency = 70;
        const double MaxFrequency = 1200;
        const double AFrequency = 440;
        static double ToneStep = Math.Pow(2, 1.0 / 12);
        static ScaleLabel[] Labels = 
        {
             new ScaleLabel() { Title = "E", Frequency =  82.4069, Color=Color.LightGreen},
             new ScaleLabel() { Title = "A", Frequency = 110.0000, Color=Color.LightGreen},
             new ScaleLabel() { Title = "D", Frequency = 146.8324, Color=Color.LightGreen},
             new ScaleLabel() { Title = "G", Frequency = 195.9977, Color=Color.LightGreen},
             new ScaleLabel() { Title = "B", Frequency = 246.9417, Color=Color.LightGreen},
             new ScaleLabel() { Title = "E", Frequency = 329.6276, Color=Color.LightGreen},
             new ScaleLabel() { Title = "",  Frequency = 440.0000, Color=Color.Silver}
        };        

        double currentFrequency;

        [DefaultValue(0.0)]
        public double CurrentFrequency
        {
            get { return currentFrequency; }
            set
            {
                if (currentFrequency != value)
                { 
                    currentFrequency = value; Invalidate(); 
                }
            }
        }

        bool signalDetected = false;

        [DefaultValue(false)]
        public bool SignalDetected
        {
            get { return signalDetected; }
            set
            {
                if (signalDetected != value)
                {
                    signalDetected = value; Invalidate();
                }
            }
        }

        public FrequenciesScale()
        {
            InitializeComponent();

            InitializeComponent2();
        }

        public FrequenciesScale(IContainer container)
        {
            container.Add(this);

            InitializeComponent();

            InitializeComponent2();
        }

        private void InitializeComponent2()
        {
            SetStyle(ControlStyles.ResizeRedraw, true);
            SetStyle(ControlStyles.OptimizedDoubleBuffer, true);
        }

        static Pen MarkerPen = new Pen(Color.Black);
        static Brush ActiveSliderBrush1 = new SolidBrush(Color.GreenYellow);
        static Brush ActiveSliderBrush2 = new SolidBrush(Color.Green);
        static Brush InactiveSliderBrush1 = new SolidBrush(Color.FromArgb(70, Color.Gray));
        static Brush InactiveSliderBrush2 = new SolidBrush(Color.FromArgb(50, Color.Black));
        const int DisplayPadding = 5;
        const int MarkWidth = 6;
        const int LabelMarkSize = 9;

        protected override void OnPaint(PaintEventArgs e)
        {
            //base.OnPaint(e);

            int minStep = (int)Math.Floor(GetToneStep(MinFrequency));
            int maxStep = (int)Math.Ceiling( GetToneStep(MaxFrequency) );

            int center = Width / 2;

            int totalSteps = maxStep - minStep;
            float stepSize = (float)(this.Height - 2 * DisplayPadding) / totalSteps;

            for (int i = 0; i <= totalSteps; i++)
            {
                float y = stepSize * i + DisplayPadding;

                e.Graphics.DrawLine(MarkerPen, center - MarkWidth / 2, y, center + MarkWidth / 2, y);
            }

            float maxTextWidth = e.Graphics.MeasureString("W", Font).Width;
            foreach (ScaleLabel label in Labels)
            {
                Brush labelBrush = new SolidBrush(label.Color);
                double labelStep = GetToneStep(label.Frequency);
                float labelPosition = (float)(stepSize * (maxStep - labelStep) + DisplayPadding);

                RectangleF labelMarkPosition = new RectangleF(DisplayPadding, labelPosition - LabelMarkSize / 2,
                    LabelMarkSize, LabelMarkSize);
                e.Graphics.FillEllipse(labelBrush, labelMarkPosition);
                e.Graphics.DrawEllipse(MarkerPen, labelMarkPosition);
                e.Graphics.FillEllipse(Brushes.White, DisplayPadding + LabelMarkSize / 5, labelPosition - LabelMarkSize / 3,
                    LabelMarkSize / 3, LabelMarkSize / 3);

                SizeF titleSize = e.Graphics.MeasureString(label.Title, Font);

                e.Graphics.DrawString(label.Title, Font, Brushes.Black,
                    new PointF(Width - DisplayPadding - maxTextWidth / 2 - titleSize.Width / 2 , 
                        labelPosition - titleSize.Height / 2));
            }

            if (CurrentFrequency > 0)
            {
                Brush sliderBrush1, sliderBrush2;
                if (!SignalDetected)
                {
                    sliderBrush1 = InactiveSliderBrush1;
                    sliderBrush2 = InactiveSliderBrush2;
                }
                else
                {
                    sliderBrush1 = ActiveSliderBrush1;
                    sliderBrush2 = ActiveSliderBrush2;
                }

                double sliderStep = GetToneStep(CurrentFrequency);
                float sliderPosition = (float)(stepSize * (maxStep - sliderStep) + DisplayPadding);

                e.Graphics.FillPolygon(sliderBrush1, new PointF[] 
                {
                    new PointF(center - 10, sliderPosition),
                    new PointF(center, sliderPosition - 5),
                    new PointF(center, sliderPosition + 5),
                    new PointF(center + 10, sliderPosition)
                });
                e.Graphics.FillPolygon(sliderBrush2, new PointF[] 
                {
                    new PointF(center - 10, sliderPosition),
                    new PointF(center, sliderPosition + 5),
                    new PointF(center, sliderPosition - 5),
                    new PointF(center + 10, sliderPosition)
                });
            }

        }

        private double GetToneStep(double frequency)
        {
            return Math.Log(frequency / AFrequency, ToneStep);
        }

        class ScaleLabel
        {
            public string Title;
            public double Frequency;
            public Color Color;
        }
    }
}

File: vb.net/FftGuitarTuner/FrequenciesScale.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Diagnostics
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Drawing

Public Partial Class FrequenciesScale
	Inherits UserControl
	Const MinFrequency As Double = 70
	Const MaxFrequency As Double = 1200
	Const AFrequency As Double = 440
	Shared ToneStep As Double = Math.Pow(2, 1.0 / 12)
	Shared Labels As ScaleLabel() = {New ScaleLabel() With { _
		.Title = "E", _
		.Frequency = 82.4069, _
		.Color = Color.LightGreen _
	}, New ScaleLabel() With { _
		.Title = "A", _
		.Frequency = 110.0, _
		.Color = Color.LightGreen _
	}, New ScaleLabel() With { _
		.Title = "D", _
		.Frequency = 146.8324, _
		.Color = Color.LightGreen _
	}, New ScaleLabel() With { _
		.Title = "G", _
		.Frequency = 195.9977, _
		.Color = Color.LightGreen _
	}, New ScaleLabel() With { _
		.Title = "B", _
		.Frequency = 246.9417, _
		.Color = Color.LightGreen _
	}, New ScaleLabel() With { _
		.Title = "E", _
		.Frequency = 329.6276, _
		.Color = Color.LightGreen _
	}, _
		New ScaleLabel() With { _
		.Title = "", _
		.Frequency = 440.0, _
		.Color = Color.Silver _
	}}

	Private m_currentFrequency As Double

	<DefaultValue(0.0)> _
	Public Property CurrentFrequency() As Double
		Get
			Return m_currentFrequency
		End Get
		Set
			If m_currentFrequency <> value Then
				m_currentFrequency = value
				Invalidate()
			End If
		End Set
	End Property

	Private m_signalDetected As Boolean = False

	<DefaultValue(False)> _
	Public Property SignalDetected() As Boolean
		Get
			Return m_signalDetected
		End Get
		Set
			If m_signalDetected <> value Then
				m_signalDetected = value
				Invalidate()
			End If
		End Set
	End Property

	Public Sub New()
		InitializeComponent()

		InitializeComponent2()
	End Sub

	Public Sub New(container As IContainer)
		container.Add(Me)

		InitializeComponent()

		InitializeComponent2()
	End Sub

	Private Sub InitializeComponent2()
		SetStyle(ControlStyles.ResizeRedraw, True)
		SetStyle(ControlStyles.OptimizedDoubleBuffer, True)
	End Sub

	Shared MarkerPen As New Pen(Color.Black)
	Shared ActiveSliderBrush1 As Brush = New SolidBrush(Color.GreenYellow)
	Shared ActiveSliderBrush2 As Brush = New SolidBrush(Color.Green)
	Shared InactiveSliderBrush1 As Brush = New SolidBrush(Color.FromArgb(70, Color.Gray))
	Shared InactiveSliderBrush2 As Brush = New SolidBrush(Color.FromArgb(50, Color.Black))
	Const DisplayPadding As Integer = 5
	Const MarkWidth As Integer = 6
	Const LabelMarkSize As Integer = 9

	Protected Overrides Sub OnPaint(e As PaintEventArgs)
		'base.OnPaint(e);

		Dim minStep As Integer = CInt(Math.Truncate(Math.Floor(GetToneStep(MinFrequency))))
		Dim maxStep As Integer = CInt(Math.Truncate(Math.Ceiling(GetToneStep(MaxFrequency))))

		Dim center As Integer = Width \ 2

		Dim totalSteps As Integer = maxStep - minStep
		Dim stepSize As Single = CSng(Me.Height - 2 * DisplayPadding) / totalSteps

		For i As Integer = 0 To totalSteps
			Dim y As Single = stepSize * i + DisplayPadding

			e.Graphics.DrawLine(MarkerPen, center - MarkWidth \ 2, y, center + MarkWidth \ 2, y)
		Next

		Dim maxTextWidth As Single = e.Graphics.MeasureString("W", Font).Width
		For Each label As ScaleLabel In Labels
			Dim labelBrush As Brush = New SolidBrush(label.Color)
			Dim labelStep As Double = GetToneStep(label.Frequency)
			Dim labelPosition As Single = CSng(stepSize * (maxStep - labelStep) + DisplayPadding)

			Dim labelMarkPosition As New RectangleF(DisplayPadding, labelPosition - LabelMarkSize \ 2, LabelMarkSize, LabelMarkSize)
			e.Graphics.FillEllipse(labelBrush, labelMarkPosition)
			e.Graphics.DrawEllipse(MarkerPen, labelMarkPosition)
			e.Graphics.FillEllipse(Brushes.White, DisplayPadding + LabelMarkSize \ 5, labelPosition - LabelMarkSize \ 3, LabelMarkSize \ 3, LabelMarkSize \ 3)

			Dim titleSize As SizeF = e.Graphics.MeasureString(label.Title, Font)

			e.Graphics.DrawString(label.Title, Font, Brushes.Black, New PointF(Width - DisplayPadding - maxTextWidth / 2 - titleSize.Width / 2, labelPosition - titleSize.Height / 2))
		Next

		If CurrentFrequency > 0 Then
			Dim sliderBrush1 As Brush, sliderBrush2 As Brush
			If Not SignalDetected Then
				sliderBrush1 = InactiveSliderBrush1
				sliderBrush2 = InactiveSliderBrush2
			Else
				sliderBrush1 = ActiveSliderBrush1
				sliderBrush2 = ActiveSliderBrush2
			End If

			Dim sliderStep As Double = GetToneStep(CurrentFrequency)
			Dim sliderPosition As Single = CSng(stepSize * (maxStep - sliderStep) + DisplayPadding)

			e.Graphics.FillPolygon(sliderBrush1, New PointF() {New PointF(center - 10, sliderPosition), New PointF(center, sliderPosition - 5), New PointF(center, sliderPosition + 5), New PointF(center + 10, sliderPosition)})
			e.Graphics.FillPolygon(sliderBrush2, New PointF() {New PointF(center - 10, sliderPosition), New PointF(center, sliderPosition + 5), New PointF(center, sliderPosition - 5), New PointF(center + 10, sliderPosition)})
		End If

	End Sub

	Private Function GetToneStep(frequency As Double) As Double
		Return Math.Log(frequency / AFrequency, ToneStep)
	End Function

	Private Class ScaleLabel
		Public Title As String
		Public Frequency As Double
		Public Color As Color
	End Class
End Class
